services/libs/integrations/src/integrations/github/processStream.ts

```diff
@@ -0,0 +1,844 @@
+      } else {
+        ctx.log.warn(
+          `Repo ${repo.name} will not be parsed. It is not available with the github token`,
+        )
+        unavailableRepos.push(repo)
+      }
+    }
+  }
+
+  // update integration settings
+  // this settings will be avaliable in next streams
+  await ctx.updateIntegrationSettings({
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    ...(ctx.integration.settings as any),
+    repos,
+    unavailableRepos,
+  })
+
+  // now it's time to start streams
+  for (const repo of repos) {
+    for (const endpoint of [
+      GithubStreamType.STARGAZERS,
+      GithubStreamType.FORKS,
+      GithubStreamType.PULLS,
+      GithubStreamType.ISSUES,
+      GithubStreamType.DISCUSSIONS,
+    ]) {
+      // this firstPage thing is important to avoid duplicate streams and for handleNextPageStream to work
+      await ctx.publishStream<GithubBasicStream>(`${endpoint}:${repo.name}:firstPage`, {
+        repo,
+        page: '',
+      })
+    }
+  }
+}
+
+const processStargazersStream: ProcessStreamHandler = async (ctx) => {
+  const data = ctx.stream.data as GithubBasicStream
+  const stargazersQuery = new StargazersQuery(data.repo, ctx.integration.token)
+  const result = await stargazersQuery.getSinglePage(data.page)
+  result.data = result.data.filter((i) => (i as any).node?.login)
+
+  // handle next page
+  await handleNextPageStream(ctx, result)
+
+  for (const record of result.data) {
+    const member = await prepareMember(record.node, ctx)
+
+    // publish data
+    await ctx.publishData<GithubApiData>({
+      type: GithubActivityType.STAR,
+      data: record,
+      member,
+      repo: data.repo,
+    })
+  }
+}
+
+const processForksStream: ProcessStreamHandler = async (ctx) => {
+  const data = ctx.stream.data as GithubBasicStream
+  const forksQuery = new ForksQuery(data.repo, ctx.integration.token)
+  const result = await forksQuery.getSinglePage(data.page)
+
+  // filter out activities without authors (such as bots) -- may not the case for forks, but filter out anyway
+  result.data = result.data.filter((i) => (i as any).owner?.login)
+
+  // handle next page
+  await handleNextPageStream(ctx, result)
+
+  for (const record of result.data) {
+    const member = await prepareMember(record.owner, ctx)
+
+    // publish data
+    await ctx.publishData<GithubApiData>({
+      type: GithubActivityType.FORK,
+      data: record,
+      member,
+      repo: data.repo,
+    })
+  }
+}
+
+const processPullsStream: ProcessStreamHandler = async (ctx) => {
+  const data = ctx.stream.data as GithubBasicStream
+  const forksQuery = new PullRequestsQuery(data.repo, ctx.integration.token)
+  const result = await forksQuery.getSinglePage(data.page)
+
+  // filter out activities without authors (such as bots)
+  result.data = result.data.filter((i) => (i as any).author?.login)
+
+  // handle next page
+  await handleNextPageStream(ctx, result)
+
+  for (const record1 of result.data) {
```

Comment:
NIT: let's give `record1` and `record2` better names, maybe `pulls` and `pullEvents`

Reply:
updated